ipatests/test_integration/test_otp.py

```diff
@@ -458,41 +458,80 @@ def test_totp_ldap(self):
         master = self.master
         basedn = master.domain.basedn
         USER1 = 'user-forced-otp'
+        TMP_PASSWORD = 'Secret1234509'
         binddn = DN(f"uid={USER1},cn=users,cn=accounts,{basedn}")
 
-        tasks.create_active_user(master, USER1, PASSWORD)
         tasks.kinit_admin(master)
+        master.run_command(['ipa', 'pwpolicy-mod', '--minlife', '0'])
+        tasks.user_add(master, USER1, password=TMP_PASSWORD)
         # Enforce use of OTP token for this user
         master.run_command(['ipa', 'user-mod', USER1,
                             '--user-auth-type=otp'])
         try:
+            # Change initial password through the IPA endpoint
+            url = f'https://{master.hostname}/ipa/session/change_password'
+            master.run_command(['curl', '-d', f'user={USER1}',
+                                '-d', f'old_password={TMP_PASSWORD}',
+                                '-d', f'new_password={PASSWORD}',
+                                '--referer', f'https://{master.hostname}/ipa',
+                                url])
             conn = master.ldap_connect()
             # First, attempt authenticating with a password but without LDAP
             # control to enforce OTP presence and without server-side
             # enforcement of the OTP presence check.
             conn.simple_bind(binddn, f"{PASSWORD}")
-            # Add an OTP token now
-            otpuid, totp = add_otptoken(master, USER1, otptype="totp")
             # Next, enforce Password+OTP for a user with OTP token
             master.run_command(['ipa', 'config-mod', '--addattr',
                                 'ipaconfigstring=EnforceLDAPOTP'])
+            # Try to bind without OTP because there is no OTP token yet,
+            # the operation should succeed because OTP enforcement is implicit
```

Comment:
I think this wording is confusing. It succeeds because enforcement is implicit AND there is no token yet. If there had been a token then the request would have failed without an OTP.
It might be useful to do an implicit auth (EnforceLDAPOTP) bind with no OTP.
As I understand it the request becomes explicit once the LDAP control is set.

It would be helpful to describe this flow in the commit.

Otherwise the change looks good.

Reply:
Thanks. I added more details in the commit message.